File: Consumer Purchase Prediction using Machine Learning/scripts/r/ml_analysis.R
# Machine Learning Analysis Script
# Consumer Purchase Prediction Project

# Load necessary libraries
library(dplyr)
library(caret)
library(randomForest)
library(e1071)
library(rpart)
library(rpart.plot)
library(pROC)
library(ROCR)

# Function to find project root by looking for data directory
find_project_root <- function() {
  current_dir <- getwd()
  max_levels <- 10
  project_marker <- file.path("Consumer Purchase Prediction", "Consumer Purchase Prediction", "data", "Advertisement.csv")
  
  for (i in 1:max_levels) {
    if (file.exists(file.path(current_dir, project_marker))) {
      return(current_dir)
    }
    if (file.exists(file.path(current_dir, "data", "Advertisement.csv"))) {
      return(current_dir)
    }
    if (basename(current_dir) == "Consumer Purchase Prediction") {
      if (file.exists(file.path(current_dir, "Consumer Purchase Prediction", "data", "Advertisement.csv"))) {
        return(current_dir)
      }
      if (file.exists(file.path(current_dir, "data", "Advertisement.csv"))) {
        return(current_dir)
      }
    }
    parent_dir <- dirname(current_dir)
    if (parent_dir == current_dir) break
    current_dir <- parent_dir
  }
  return(NULL)
}

# Set working directory to project root
project_root <- find_project_root()
if (!is.null(project_root)) {
  setwd(project_root)
  cat("Working directory set to:", getwd(), "\n")
} else {
  cat("Warning: Could not find project root. Using current directory:", getwd(), "\n")
}

# Load the dataset - try multiple possible paths
data_paths <- c(
  file.path("Consumer Purchase Prediction", "Consumer Purchase Prediction", "data", "Advertisement.csv"),
  file.path("data", "Advertisement.csv"),
  "Advertisement.csv"
)

data_path <- NULL
for (path in data_paths) {
  if (file.exists(path)) {
    data_path <- path
    break
  }
}

if (is.null(data_path)) {
  stop(paste("Cannot find Advertisement.csv. Searched in:\n",
             paste("  -", data_paths, collapse = "\n"),
             "\nCurrent working directory:", getwd()))
}

df <- read.csv(data_path, stringsAsFactors = TRUE)

cat("Dataset loaded successfully from:", data_path, "\n")
cat("MACHINE LEARNING ANALYSIS\n")
cat(paste(rep("=", 50), collapse=""), "\n\n")

# Preprocessing
df$Gender <- as.factor(df$Gender)
df$Purchased <- as.factor(df$Purchased)

# Prepare features and target
X <- df[, c("Gender", "Age", "EstimatedSalary")]
y <- df$Purchased

# Split the data
set.seed(42)
trainIndex <- createDataPartition(y, p = 0.8, list = FALSE)
X_train <- X[trainIndex, ]
X_test <- X[-trainIndex, ]
y_train <- y[trainIndex]
y_test <- y[-trainIndex]

cat("Training set size:", nrow(X_train), "\n")
cat("Test set size:", nrow(X_test), "\n\n")

# Create training data frame
train_data <- data.frame(X_train, Purchased = y_train)
test_data <- data.frame(X_test, Purchased = y_test)

# 1. Logistic Regression
cat("1. LOGISTIC REGRESSION\n")
cat(paste(rep("-", 50), collapse=""), "\n")
logistic_model <- train(Purchased ~ ., data = train_data, method = "glm", 
                        family = "binomial", trControl = trainControl(method = "cv", number = 5))
logistic_pred <- predict(logistic_model, test_data)
logistic_pred_proba <- predict(logistic_model, test_data, type = "prob")[, 2]
logistic_cm <- confusionMatrix(logistic_pred, y_test)
cat("Accuracy:", logistic_cm$overall["Accuracy"], "\n")
cat("Precision:", logistic_cm$byClass["Precision"], "\n")
cat("Recall:", logistic_cm$byClass["Recall"], "\n")
cat("F1 Score:", logistic_cm$byClass["F1"], "\n\n")

# 2. Random Forest
cat("2. RANDOM FOREST\n")
cat(paste(rep("-", 50), collapse=""), "\n")
rf_model <- train(Purchased ~ ., data = train_data, method = "rf", 
                  ntree = 100, trControl = trainControl(method = "cv", number = 5))
rf_pred <- predict(rf_model, test_data)
rf_pred_proba <- predict(rf_model, test_data, type = "prob")[, 2]
rf_cm <- confusionMatrix(rf_pred, y_test)
cat("Accuracy:", rf_cm$overall["Accuracy"], "\n")
cat("Precision:", rf_cm$byClass["Precision"], "\n")
cat("Recall:", rf_cm$byClass["Recall"], "\n")
cat("F1 Score:", rf_cm$byClass["F1"], "\n\n")

# 3. SVM
cat("3. SUPPORT VECTOR MACHINE\n")
cat(paste(rep("-", 50), collapse=""), "\n")
svm_model <- train(Purchased ~ ., data = train_data, method = "svmRadial", 
                   trControl = trainControl(method = "cv", number = 5))
svm_pred <- predict(svm_model, test_data)
svm_pred_proba <- predict(svm_model, test_data, type = "prob")[, 2]
svm_cm <- confusionMatrix(svm_pred, y_test)
cat("Accuracy:", svm_cm$overall["Accuracy"], "\n")
cat("Precision:", svm_cm$byClass["Precision"], "\n")
cat("Recall:", svm_cm$byClass["Recall"], "\n")
cat("F1 Score:", svm_cm$byClass["F1"], "\n\n")

# 4. Naive Bayes
cat("4. NAIVE BAYES\n")
cat(paste(rep("-", 50), collapse=""), "\n")
nb_model <- train(Purchased ~ ., data = train_data, method = "nb", 
                  trControl = trainControl(method = "cv", number = 5))
nb_pred <- predict(nb_model, test_data)
nb_pred_proba <- predict(nb_model, test_data, type = "prob")[, 2]
nb_cm <- confusionMatrix(nb_pred, y_test)
cat("Accuracy:", nb_cm$overall["Accuracy"], "\n")
cat("Precision:", nb_cm$byClass["Precision"], "\n")
cat("Recall:", nb_cm$byClass["Recall"], "\n")
cat("F1 Score:", nb_cm$byClass["F1"], "\n\n")

# 5. Decision Tree
cat("5. DECISION TREE\n")
cat(paste(rep("-", 50), collapse=""), "\n")
dt_model <- train(Purchased ~ ., data = train_data, method = "rpart", 
                  trControl = trainControl(method = "cv", number = 5))
dt_pred <- predict(dt_model, test_data)
dt_pred_proba <- predict(dt_model, test_data, type = "prob")[, 2]
dt_cm <- confusionMatrix(dt_pred, y_test)
cat("Accuracy:", dt_cm$overall["Accuracy"], "\n")
cat("Precision:", dt_cm$byClass["Precision"], "\n")
cat("Recall:", dt_cm$byClass["Recall"], "\n")
cat("F1 Score:", dt_cm$byClass["F1"], "\n\n")

# Model Comparison
cat("MODEL COMPARISON\n")
cat(paste(rep("=", 50), collapse=""), "\n\n")

results <- data.frame(
  Model = c("Logistic Regression", "Random Forest", "SVM", "Naive Bayes", "Decision Tree"),
  Accuracy = c(logistic_cm$overall["Accuracy"], rf_cm$overall["Accuracy"], 
               svm_cm$overall["Accuracy"], nb_cm$overall["Accuracy"], dt_cm$overall["Accuracy"]),
  Precision = c(logistic_cm$byClass["Precision"], rf_cm$byClass["Precision"], 
                svm_cm$byClass["Precision"], nb_cm$byClass["Precision"], dt_cm$byClass["Precision"]),
  Recall = c(logistic_cm$byClass["Recall"], rf_cm$byClass["Recall"], 
             svm_cm$byClass["Recall"], nb_cm$byClass["Recall"], dt_cm$byClass["Recall"]),
  F1 = c(logistic_cm$byClass["F1"], rf_cm$byClass["F1"], 
         svm_cm$byClass["F1"], nb_cm$byClass["F1"], dt_cm$byClass["F1"])
)

results <- results[order(-results$Accuracy), ]
print(results)

# Find best model
best_model_idx <- which.max(results$Accuracy)
best_model_name <- results$Model[best_model_idx]

cat("\nBEST MODEL:", best_model_name, "\n")
cat(paste(rep("=", 50), collapse=""), "\n")
cat("Accuracy:", results$Accuracy[best_model_idx], "\n")
cat("Precision:", results$Precision[best_model_idx], "\n")
cat("Recall:", results$Recall[best_model_idx], "\n")
cat("F1 Score:", results$F1[best_model_idx], "\n")

# Get output directory
output_paths <- c(
  file.path("Consumer Purchase Prediction", "Consumer Purchase Prediction", "output"),
  "output"
)

output_dir <- NULL
for (path in output_paths) {
  if (dir.exists(path)) {
    output_dir <- path
    break
  }
}

if (is.null(output_dir)) {
  output_dir <- output_paths[1]
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
}

# ROC Curves
png(file.path(output_dir, "roc_curves_r.png"), width = 1000, height = 800, res = 300)
roc_logistic <- roc(y_test, logistic_pred_proba)
roc_rf <- roc(y_test, rf_pred_proba)
roc_svm <- roc(y_test, svm_pred_proba)
roc_nb <- roc(y_test, nb_pred_proba)
roc_dt <- roc(y_test, dt_pred_proba)

plot(roc_logistic, col = "blue", main = "ROC Curves for All Models")
lines(roc_rf, col = "red")
lines(roc_svm, col = "green")
lines(roc_nb, col = "purple")
lines(roc_dt, col = "orange")
legend("bottomright", 
       legend = c(paste("Logistic Regression (AUC =", round(auc(roc_logistic), 3), ")"),
                  paste("Random Forest (AUC =", round(auc(roc_rf), 3), ")"),
                  paste("SVM (AUC =", round(auc(roc_svm), 3), ")"),
                  paste("Naive Bayes (AUC =", round(auc(roc_nb), 3), ")"),
                  paste("Decision Tree (AUC =", round(auc(roc_dt), 3), ")")),
       col = c("blue", "red", "green", "purple", "orange"), lty = 1)
dev.off()

cat("\n", paste(rep("=", 50), collapse=""), "\n")
cat("ML ANALYSIS COMPLETED SUCCESSFULLY!\n")
cat(paste(rep("=", 50), collapse=""), "\n")

File: Consumer Purchase Prediction using Machine Learning/scripts/python/ml_analysis.py
"""
Machine Learning Analysis Script
Consumer Purchase Prediction Project
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, confusion_matrix, roc_curve, auc
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
import warnings
import pickle
import os
warnings.filterwarnings('ignore')

def load_and_preprocess_data(file_path):
    """Load and preprocess data"""
    df = pd.read_csv(file_path)
    
    # Encode categorical variables
    le = LabelEncoder()
    df['Gender'] = le.fit_transform(df['Gender'])
    
    # Prepare features and target
    X = df[['Gender', 'Age', 'EstimatedSalary']]
    y = df['Purchased']
    
    # Split the data
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=y
    )
    
    # Scale the features
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)
    
    return X_train, X_test, X_train_scaled, X_test_scaled, y_train, y_test, scaler, le

def train_and_evaluate_models(X_train, X_test, X_train_scaled, X_test_scaled, y_train, y_test):
    """Train and evaluate multiple models"""
    models = {
        'Logistic Regression': LogisticRegression(random_state=42, max_iter=1000),
        'Random Forest': RandomForestClassifier(n_estimators=100, random_state=42),
        'SVM': SVC(probability=True, random_state=42),
        'Naive Bayes': GaussianNB(),
        'Decision Tree': DecisionTreeClassifier(random_state=42),
        'Gradient Boosting': GradientBoostingClassifier(random_state=42)
    }
    
    results = {}
    
    for name, model in models.items():
        # Train model
        if name == 'Logistic Regression' or name == 'SVM':
            model.fit(X_train_scaled, y_train)
            y_pred = model.predict(X_test_scaled)
            y_pred_proba = model.predict_proba(X_test_scaled)[:, 1]
        else:
            model.fit(X_train, y_train)
            y_pred = model.predict(X_test)
            y_pred_proba = model.predict_proba(X_test)[:, 1]
        
        # Calculate metrics
        accuracy = accuracy_score(y_test, y_pred)
        precision = precision_score(y_test, y_pred)
        recall = recall_score(y_test, y_pred)
        f1 = f1_score(y_test, y_pred)
        
        # Cross-validation score
        if name == 'Logistic Regression' or name == 'SVM':
            cv_scores = cross_val_score(model, X_train_scaled, y_train, cv=5, scoring='accuracy')
        else:
            cv_scores = cross_val_score(model, X_train, y_train, cv=5, scoring='accuracy')
        
        results[name] = {
            'model': model,
            'accuracy': accuracy,
            'precision': precision,
            'recall': recall,
            'f1': f1,
            'cv_mean': cv_scores.mean(),
            'cv_std': cv_scores.std(),
            'y_pred': y_pred,
            'y_pred_proba': y_pred_proba
        }
        
        print(f"\n{name}:")
        print(f"  Accuracy: {accuracy:.4f}")
        print(f"  Precision: {precision:.4f}")
        print(f"  Recall: {recall:.4f}")
        print(f"  F1 Score: {f1:.4f}")
        print(f"  CV Accuracy: {cv_scores.mean():.4f} (+/- {cv_scores.std()*2:.4f})")
    
    return results

def plot_results(results, y_test, output_dir='../../output'):
    """Plot model comparison and evaluation metrics"""
    os.makedirs(output_dir, exist_ok=True)
    
    # Model comparison
    comparison_df = pd.DataFrame({
        'Model': list(results.keys()),
        'Accuracy': [results[m]['accuracy'] for m in results.keys()],
        'Precision': [results[m]['precision'] for m in results.keys()],
        'Recall': [results[m]['recall'] for m in results.keys()],
        'F1 Score': [results[m]['f1'] for m in results.keys()],
        'CV Accuracy': [results[m]['cv_mean'] for m in results.keys()]
    })
    
    comparison_df = comparison_df.sort_values('Accuracy', ascending=False)
    print("\nModel Comparison:")
    print(comparison_df.to_string(index=False))
    
    # Visualization
    fig, ax = plt.subplots(figsize=(12, 6))
    x = np.arange(len(comparison_df))
    width = 0.15
    
    ax.bar(x - 2*width, comparison_df['Accuracy'], width, label='Accuracy')
    ax.bar(x - width, comparison_df['Precision'], width, label='Precision')
    ax.bar(x, comparison_df['Recall'], width, label='Recall')
    ax.bar(x + width, comparison_df['F1 Score'], width, label='F1 Score')
    ax.bar(x + 2*width, comparison_df['CV Accuracy'], width, label='CV Accuracy')
    
    ax.set_xlabel('Models')
    ax.set_ylabel('Score')
    ax.set_title('Model Performance Comparison')
    ax.set_xticks(x)
    ax.set_xticklabels(comparison_df['Model'], rotation=45, ha='right')
    ax.legend()
    ax.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(f'{output_dir}/model_comparison.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # Confusion matrices
    fig, axes = plt.subplots(2, 3, figsize=(18, 12))
    axes = axes.ravel()
    
    for idx, (name, result) in enumerate(results.items()):
        cm = confusion_matrix(y_test, result['y_pred'])
        sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', ax=axes[idx])
        axes[idx].set_title(f'{name}\nAccuracy: {result["accuracy"]:.3f}')
        axes[idx].set_ylabel('True Label')
        axes[idx].set_xlabel('Predicted Label')
    
    plt.tight_layout()
    plt.savefig(f'{output_dir}/confusion_matrices.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # ROC curves
    plt.figure(figsize=(10, 8))
    for name, result in results.items():
        fpr, tpr, _ = roc_curve(y_test, result['y_pred_proba'])
        roc_auc = auc(fpr, tpr)
        plt.plot(fpr, tpr, label=f'{name} (AUC = {roc_auc:.3f})')
    
    plt.plot([0, 1], [0, 1], 'k--', label='Random Classifier')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title('ROC Curves for All Models')
    plt.legend(loc="lower right")
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(f'{output_dir}/roc_curves.png', dpi=300, bbox_inches='tight')
    plt.close()

def main():
    """Main function"""
    print("MACHINE LEARNING ANALYSIS")
    print("="*50)
    
    # Load and preprocess data
    X_train, X_test, X_train_scaled, X_test_scaled, y_train, y_test, scaler, le = \
        load_and_preprocess_data('../../data/Advertisement.csv')
    
    # Train and evaluate models
    results = train_and_evaluate_models(
        X_train, X_test, X_train_scaled, X_test_scaled, y_train, y_test
    )
    
    # Plot results
    plot_results(results, y_test)
    
    # Find best model
    best_model_name = max(results.keys(), key=lambda x: results[x]['accuracy'])
    best_model = results[best_model_name]
    
    print(f"\nBEST MODEL: {best_model_name}")
    print("="*50)
    print(f"Accuracy: {best_model['accuracy']:.4f}")
    print(f"Precision: {best_model['precision']:.4f}")
    print(f"Recall: {best_model['recall']:.4f}")
    print(f"F1 Score: {best_model['f1']:.4f}")
    print(f"\nClassification Report:")
    print(classification_report(y_test, best_model['y_pred']))
    
    # Save best model
    os.makedirs('../../models', exist_ok=True)
    with open('../../models/best_model.pkl', 'wb') as f:
        pickle.dump(best_model['model'], f)
    print(f"\nBest model saved to models/best_model.pkl")
    
    print("\n" + "="*50)
    print("ML ANALYSIS COMPLETED SUCCESSFULLY!")
    print("="*50)

if __name__ == "__main__":
    main()

